crates/jit-icache-coherence/src/libc.rs

```diff
@@ -1,86 +1,106 @@
+        //
+        // I couldn't find the follow up for that patch but there doesn't seem to be disagreement
+        // about that specific part in the replies.
+        // TODO: Check if the kernel has updated the membarrier documentation
+        //
+        // See the following issues for more info:
+        //  * https://github.com/bytecodealliance/wasmtime/pull/3426
+        //  * https://github.com/bytecodealliance/wasmtime/pull/4997
+        //
+        // TODO: x86 and s390x have coherent caches so they don't need this, but RISCV does not
+        // guarantee that, so we may need to do something similar for it. However as noted in the
+        // above kernel patch the SYNC_CORE membarrier has different guarantees on each
+        // architecture so we need follow up and check what it provides us.
+        // See: https://github.com/bytecodealliance/wasmtime/issues/5033
+        match membarrier(MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE) {
+            Ok(_) => {}
+
+            // EPERM happens if the calling process hasn't yet called the register membarrier.
+            // We can call the register membarrier now, and then retry the actual membarrier,
+            //
+            // This does have some overhead since on the first time we call this function we
+            // actually execute three membarriers, but this only happens once per process and only
+            // one slow membarrier is actually executed (The last one, which actually generates an
+            // IPI).
+            Err(e) if e.raw_os_error().unwrap() == EPERM => {
+                membarrier(MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE)?;
+                membarrier(MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE)?;
+            }
 
-        // On kernels older than 4.16 the above syscall does not exist, so we can
-        // fallback to MEMBARRIER_CMD_GLOBAL which is an alias for MEMBARRIER_CMD_SHARED
-        // that has existed since 4.3. GLOBAL is a lot slower, but allows us to have
-        // compatibility with older kernels.
-        Err(e) if e.raw_os_error().unwrap() == EINVAL => {
-            membarrier(MEMBARRIER_CMD_GLOBAL)?;
+            // On kernels older than 4.16 the above syscall does not exist, so we can
+            // fallback to MEMBARRIER_CMD_GLOBAL which is an alias for MEMBARRIER_CMD_SHARED
+            // that has existed since 4.3. GLOBAL is a lot slower, but allows us to have
+            // compatibility with older kernels.
+            Err(e) if e.raw_os_error().unwrap() == EINVAL => {
+                membarrier(MEMBARRIER_CMD_GLOBAL)?;
+            }
+
+            // In any other case we got an actual error, so lets propagate that up
+            e => e?,
         }
 
-        // In any other case we got an actual error, so lets propagate that up
-        e => e?,
+        Ok(())
     }
 
-    Ok(())
+    fn membarrier(barrier: libc::c_int) -> Result<()> {
```

Comment:
It looks like `membarrier` is now defined only on AArch64, whereas previously it was defined on all architectures (and IIRC it's used on RISC-V, or at least will eventually need to be, as well) -- could you revert this bit of code-motion?

Reply:
RISC-V is going to need a custom syscall for this operation. IIRC the kernel membarrier interface is not a good fit for the RISC-V model.

See #5033 and #5078 for more details.

Moving this out may still be useful for other arch's though.

Reply:
Fine, I can revert this change. I try to favor the current state of things over future possible uses, but I don't have a strong opinion here.

Reply:
Actually, let me have a semi-strong opinion here: if I put the function back where it was, then it needs to be either annotated with the exact same `cfg`, or with an `#[allow(unused)]` directive so the compiler doesn't complain about it being unused... So overall it seems better to keep it here, to reflect the current state?

Reply:
Ah, yeah, given the above it seems it's actually only needed on aarch64 right now, so current state is fine. Thanks :-)